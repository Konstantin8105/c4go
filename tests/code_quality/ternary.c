#define MaxMinTest(type) \
{ \
	diag(#type);       \
	type a = 54;       \
	type b = -4;       \
	type c;            \ 
	c = a > b ? a : b ;\
	c = a < b ? a : b; \
	c = b < a ? a : b ;\
	c = b > a ? a : b; \
	                   \
	c = a > b ? a : b + 1 ;\
	c = a < b ? a + 1 : b; \
	c = b < a ? a : b + 1 ;\
	c = b > a ? a + 1 : b; \
} 

void test_min_max()
{
	MaxMinTest(char);
	MaxMinTest(short);
	MaxMinTest(int);
	MaxMinTest(float);
	MaxMinTest(double);
	MaxMinTest(long double);
}

