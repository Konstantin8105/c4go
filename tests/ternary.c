#include "tests.h"
#include <stdio.h>

void f_empty()
{
    return;
};

int ret_two(int a)
{
    return 2 + a;
}

int ret_s(int a)
{
    return a > 0;
}

int ret_ter(int a)
{
    int (*v)(int);
    v = ret_s;
    return v ? 1 : ret_two(0);
}

void test_return_ternary()
{
    is_eq(ret_ter(10), 1);
    is_eq(ret_ter(-1), 1);
}

int main()
{
    plan(13);

    int a = 'a' == 65 ? 10 : 100;
    float b = 10 == 10 ? 1.0 : 2.0;
    char* c = 'x' == 5 ? "one" : "two";
    char d = a == 100 ? 'x' : 1;

    is_eq(a, 100);
    is_eq(b, 1);
    is_streq(c, "two");
    is_eq(d, 'x');

    is_false(0 ? 1 : 0);
    is_false(NULL ? 1 : 0);
    is_true('x' ? 1 : 0);

    a = a == 10 ? b == 1.0 ? 1 : 2 : 2;

    if (a == (a == 2 ? 5 : 10)) {
        fail(__func__);
    } else {
        pass(__func__);
    }

    diag("CStyleCast <ToVoid>");
    {
        double a, b;
        0 ? (void)(a) : (void)(b);
        (void)(a), (void)(b);
    }
    {
        double a;
        0 ? (void)(a) : f_empty();
        (void)(a);
    }
    {
        double b;
        0 ? f_empty() : (void)(b);
        (void)(b);
    }
    {
        ;
        0 ? f_empty() : f_empty();
    }
    pass("Ok - ToVoid");
    {
        diag("oper ++");
        int b = 42;
        int addr = 0;
        b = addr++ ? 1 : 2;
        is_eq(addr, 1);
        is_eq(b, 2);
    }

    test_return_ternary();

    done_testing();
}
