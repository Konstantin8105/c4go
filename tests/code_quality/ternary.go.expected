//
//	Package - transpiled by c4go
//
//	If you have found any issues, please raise an issue at:
//	https://github.com/Konstantin8105/c4go/
//

package code_quality

// test_min_max - transpiled function from  C4GO/tests/code_quality/ternary.c:18
func test_min_max() {
	{
		diag([]byte("char\x00"))
		var a byte = byte(54)
		var b byte = byte(-4 + 256)
		var c byte
		c = byte(func() int32 {
			if int32(a) > int32(b) {
				return int32(a)
			}
			return int32(b)
		}())
		c = byte(func() int32 {
			if int32(a) < int32(b) {
				return int32(a)
			}
			return int32(b)
		}())
		c = byte(func() int32 {
			if int32(b) < int32(a) {
				return int32(a)
			}
			return int32(b)
		}())
		c = byte(func() int32 {
			if int32(b) > int32(a) {
				return int32(a)
			}
			return int32(b)
		}())
		c = byte(func() int32 {
			if int32(a) > int32(b) {
				return int32(a)
			}
			return int32(b) + 1
		}())
		c = byte(func() int32 {
			if int32(a) < int32(b) {
				return int32(a) + 1
			}
			return int32(b)
		}())
		c = byte(func() int32 {
			if int32(b) < int32(a) {
				return int32(a)
			}
			return int32(b) + 1
		}())
		c = byte(func() int32 {
			if int32(b) > int32(a) {
				return int32(a) + 1
			}
			return int32(b)
		}())
	}
	{
		diag([]byte("short\x00"))
		var a int16 = 54
		var b int16 = int16(-4)
		var c int16
		c = int16(func() int32 {
			if int32(a) > int32(b) {
				return int32(a)
			}
			return int32(b)
		}())
		c = int16(func() int32 {
			if int32(a) < int32(b) {
				return int32(a)
			}
			return int32(b)
		}())
		c = int16(func() int32 {
			if int32(b) < int32(a) {
				return int32(a)
			}
			return int32(b)
		}())
		c = int16(func() int32 {
			if int32(b) > int32(a) {
				return int32(a)
			}
			return int32(b)
		}())
		c = int16(func() int32 {
			if int32(a) > int32(b) {
				return int32(a)
			}
			return int32(b) + 1
		}())
		c = int16(func() int32 {
			if int32(a) < int32(b) {
				return int32(a) + 1
			}
			return int32(b)
		}())
		c = int16(func() int32 {
			if int32(b) < int32(a) {
				return int32(a)
			}
			return int32(b) + 1
		}())
		c = int16(func() int32 {
			if int32(b) > int32(a) {
				return int32(a) + 1
			}
			return int32(b)
		}())
	}
	{
		diag([]byte("int\x00"))
		var a int32 = 54
		var b int32 = -4
		var c int32
		if a > b {
			c = a
		} else {
			c = b
		}
		if a < b {
			c = a
		} else {
			c = b
		}
		if b < a {
			c = a
		} else {
			c = b
		}
		if b > a {
			c = a
		} else {
			c = b
		}
		if a > b {
			c = a
		} else {
			c = b + 1
		}
		if a < b {
			c = a + 1
		} else {
			c = b
		}
		if b < a {
			c = a
		} else {
			c = b + 1
		}
		if b > a {
			c = a + 1
		} else {
			c = b
		}
	}
	{
		diag([]byte("float\x00"))
		var a float32 = 54
		var b float32 = float32(-4)
		var c float32
		if a > b {
			c = a
		} else {
			c = b
		}
		if a < b {
			c = a
		} else {
			c = b
		}
		if b < a {
			c = a
		} else {
			c = b
		}
		if b > a {
			c = a
		} else {
			c = b
		}
		if a > b {
			c = a
		} else {
			c = b + 1
		}
		if a < b {
			c = a + 1
		} else {
			c = b
		}
		if b < a {
			c = a
		} else {
			c = b + 1
		}
		if b > a {
			c = a + 1
		} else {
			c = b
		}
	}
	{
		diag([]byte("double\x00"))
		var a float64 = 54
		var b float64 = float64(-4)
		var c float64
		if a > b {
			c = a
		} else {
			c = b
		}
		if a < b {
			c = a
		} else {
			c = b
		}
		if b < a {
			c = a
		} else {
			c = b
		}
		if b > a {
			c = a
		} else {
			c = b
		}
		if a > b {
			c = a
		} else {
			c = b + 1
		}
		if a < b {
			c = a + 1
		} else {
			c = b
		}
		if b < a {
			c = a
		} else {
			c = b + 1
		}
		if b > a {
			c = a + 1
		} else {
			c = b
		}
	}
	{
		diag([]byte("long double\x00"))
		var a float64 = 54
		var b float64 = float64(-4)
		var c float64
		if a > b {
			c = a
		} else {
			c = b
		}
		if a < b {
			c = a
		} else {
			c = b
		}
		if b < a {
			c = a
		} else {
			c = b
		}
		if b > a {
			c = a
		} else {
			c = b
		}
		if a > b {
			c = a
		} else {
			c = b + 1
		}
		if a < b {
			c = a + 1
		} else {
			c = b
		}
		if b < a {
			c = a
		} else {
			c = b + 1
		}
		if b > a {
			c = a + 1
		} else {
			c = b
		}
	}
}
